File: rust/bron_kerbosch/src/bron_kerbosch2_rp.rs
//! Bron-Kerbosch algorithm with pivot picked randomly (IK_RP)

use crate::bron_kerbosch_pivot::{visit, PivotChoice};
use crate::graph::{connected_vertices, UndirectedGraph, VertexSetLike};
use crate::reporter::Reporter;

pub fn explore<Graph, Rprtr>(graph: &Graph, reporter: &mut Rprtr)
where
    Graph: UndirectedGraph,
    Rprtr: Reporter,
{
    let candidates = connected_vertices(graph);
    if !candidates.is_empty() {
        visit(
            graph,
            reporter,
            PivotChoice::Random,
            PivotChoice::Random,
            candidates,
            Graph::VertexSet::new(),
            None,
        );
    }
}

File: python3/bron_kerbosch2_rp.py
# coding: utf-8

from bron_kerbosch_pivot import pick_random, visit
from graph import UndirectedGraph
from reporter import Reporter


def bron_kerbosch2_rp(graph: UndirectedGraph, reporter: Reporter):
    '''Bron-Kerbosch algorithm with pivot chosen randomly (IK_RP)'''
    if candidates := graph.connected_vertices():
        visit(graph=graph,
              reporter=reporter,
              initial_pivot_choice=pick_random,
              further_pivot_choice=pick_random,
              candidates=candidates,
              excluded=set(),
              clique=[])
